src/rust/storage/seg/src/hashtable/mod.rs

```diff
@@ -109,6 +109,131 @@ pub(crate) struct HashTable {
     next_to_chain: u64,
 }
 
+struct Iter<'a> {
+    hashtable: &'a HashTable,
+    bucket_id: usize,
+    item_slot: usize,
+    chain_len: usize,
+    chain_idx: usize,
+    finished: bool,
+}
+
+struct IterMut<'a> {
+    ptr: *mut HashBucket,
+    bucket_id: usize,
+    item_slot: usize,
+    chain_len: usize,
+    chain_idx: usize,
+    finished: bool,
+    _marker: std::marker::PhantomData<&'a mut u64>,
+}
+
+impl<'a> Iter<'a> {
+    fn new(hashtable: &'a HashTable, hash: u64) -> Self {
+        let bucket_id = (hash & hashtable.mask) as usize;
+        let bucket = hashtable.data[bucket_id];
+        let chain_len = chain_len(bucket.data[0]) as usize;
+
+        Self {
+            hashtable,
+            bucket_id,
+            item_slot: 1,
+            chain_len,
+            chain_idx: 0,
+            finished: false,
+        }
+    }
+}
+
+impl<'a> IterMut<'a> {
+    fn new(hashtable: &'a mut HashTable, hash: u64) -> Self {
+        let bucket_id = (hash & hashtable.mask) as usize;
+        let bucket = hashtable.data[bucket_id];
+        let chain_len = chain_len(bucket.data[0]) as usize;
+
+        let ptr = hashtable.data.as_mut_ptr();
+
+        Self {
+            ptr,
+            bucket_id,
+            item_slot: 1,
+            chain_len,
+            chain_idx: 0,
+            finished: false,
+            _marker: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<'a> Iterator for Iter<'a> {
+    type Item = &'a u64;
+
+    fn next(&mut self) -> Option<<Self as Iterator>::Item> {
+        if self.finished {
+            return None;
+        }
+
+        let n_item_slot = if self.chain_idx == self.chain_len {
+            N_BUCKET_SLOT
+        } else {
+            N_BUCKET_SLOT - 1
+        };
+
+        let bucket = &self.hashtable.data[self.bucket_id];
+
+        let item_info = &bucket.data[self.item_slot];
+
+        if self.item_slot < n_item_slot - 1 {
+            self.item_slot += 1;
+        } else {
+            // finished iterating in this bucket, see if it's chained
+            if self.chain_idx < self.chain_len {
+                self.chain_idx += 1;
+                self.item_slot = 0;
+                self.bucket_id = bucket.data[N_BUCKET_SLOT - 1] as usize;
+            } else {
+                self.finished = true;
+            }
+        }
+
+        Some(item_info)
+    }
+}
+
+impl<'a> Iterator for IterMut<'a> {
+    type Item = &'a mut u64;
+
+    fn next(&mut self) -> Option<<Self as Iterator>::Item> {
+        if self.finished {
+            return None;
+        }
+
+        let n_item_slot = if self.chain_idx == self.chain_len {
+            N_BUCKET_SLOT
+        } else {
+            N_BUCKET_SLOT - 1
+        };
+
+        let item_info = unsafe { &mut (*self.ptr.add(self.bucket_id)).data[self.item_slot] };
```

Comment:
@brayniac was curious about race conditions w/ HashTable and whether there's some kind of concurrency safe machinery that makes use of unsafe here sensible. Is that accomplished on some higher level via the CAS-related functions I saw elsewhere in this file or through single-threaded execution or something? 

I could probably do better trying to answer this question for myself but I lack familiarity with Segcache

Reply:
The implementation is not designed for concurrent access. We have the storage owned by a single thread. The cas function here allows the memcache cas command to be implemented and doesn't imply concurrency within the server. So it's really just cas for the clients and execution order is serialized in process.